pkg/ccl/streamingccl/logical/logical_replication_writer_processor.go

```diff
@@ -652,39 +658,57 @@ func (lrw *logicalReplicationWriterProcessor) flushBuffer(
 		return b.checkpoint, nil
 	}
 
+	batchSize := int(flushBatchSize.Get(&lrw.EvalCtx.Settings.SV))
+
 	// Ensure the batcher is always reset, even on early error returns.
 	preFlushTime := timeutil.Now()
 	lrw.debug.RecordFlushStart(preFlushTime, int64(len(b.buffer.curKVBatch)))
 
+	workers := min(maxWriterWorkers, (len(b.buffer.curKVBatch)/batchSize)+1)
+	perWorker := (len(b.buffer.curKVBatch) / workers) + 1
 	// TODO: The batching here in production would need to be much
 	// smarter. Namely, we don't want to include updates to the
 	// same key in the same batch. Also, it's possible batching
 	// will make things much worse in practice.
-	flushByteSize := 0
-	batchStart := 0
-	batchSize := int(flushBatchSize.Get(&lrw.EvalCtx.Settings.SV))
-	batchEnd := min(batchStart+batchSize, len(b.buffer.curKVBatch))
-	for batchStart < len(b.buffer.curKVBatch) && batchEnd != 0 {
-		preBatchTime := timeutil.Now()
-		batchStats, err := lrw.bh.HandleBatch(ctx, b.buffer.curKVBatch[batchStart:batchEnd])
-		if err != nil {
-			// TODO(ssd): Handle errors. We should perhaps split the batch and retry a portion of the batch.
-			// If that fails, send the failed application to the dead-letter-queue.
-			return nil, err
+	var flushByteSize atomic.Int64
+
+	g := ctxgroup.WithContext(ctx)
+	for i := 0; i < workers; i++ {
+		bh := lrw.bh[i]
+		batchStart := i * perWorker
```

Comment:
once we add timestamp and prefix sorting, we really ought to unit test these batch pre processing steps.

Reply:
We should, but I don't think it takes priority over being demo-ready right now?

I think it'll be easier once we rip out the buffer to do more component based testing too.